server/src/server.ts

```diff
@@ -54,29 +56,38 @@ const FULL_RANGE = {
  * @param {TextDocument} textDocument the textDocument we are processing
  * @returns {string} the root file path
  */
-function getTemplateRoot(pathStr, textDocument, diagnosticMap) {
+async function getTemplateRoot(pathStr, textDocument, diagnosticMap) {
 
-    let currentPath = pathStr;
+    let packageJsonPath = await findUp('package.json').then(ret_value => {

```

Comment:
Promises OR async/await - not sure why you need both.

Reply:
The bug is that If you try to run the extension on a stand-alone .ergo file it freezes up. I was trying to experiment with a hover feature which was not working at the time which was stranged. Then I noticed that the whole extension freezes up and after an extended duration of time this error comes up 

`FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10a2ea9e5 node::Abort() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 2: 0x10a2eaa54 node::Abort() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 3: 0x105ff8f77 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 4: 0x105ff8f17 v8::internal::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 5: 0x10603dde5 v8::internal::Heap::StartIdleIncrementalMarking(v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 6: 0x10603f459 v8::internal::Heap::StartIdleIncrementalMarking(v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 7: 0x10603c12c v8::internal::Heap::CreateFillerObjectAt(unsigned long, int, v8::internal::ClearRecordedSlots, v8::internal::ClearFreedMemoryMode) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 8: 0x106039d42 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
 9: 0x1060451aa v8::internal::Heap::PromotedExternalMemorySize() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
10: 0x106045591 v8::internal::Heap::PromotedExternalMemorySize() [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
11: 0x10624649b v8::internal::Factory::AllocateRawWithAllocationSite(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationType, v8::internal::Handle<v8::internal::AllocationSite>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
12: 0x10624a176 v8::internal::Factory::NewJSObjectFromMap(v8::internal::Handle<v8::internal::Map>, v8::internal::AllocationType, v8::internal::Handle<v8::internal::AllocationSite>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
13: 0x1063d3176 v8::internal::JSObject::New(v8::internal::Handle<v8::internal::JSFunction>, v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::Handle<v8::internal::AllocationSite>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
14: 0x10621e9c6 v8::internal::Factory::NewTypeError(v8::internal::MessageTemplate, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
15: 0x106166fbf v8::internal::Accessors::MakeAccessor(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Name>, void (*)(v8::Local<v8::Name>, v8::PropertyCallbackInfo<v8::Value> const&), void (*)(v8::Local<v8::Name>, v8::Local<v8::Value>, v8::PropertyCallbackInfo<v8::Boolean> const&)) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
16: 0x106b11639 v8::internal::compiler::ZoneStats::ReturnZone(v8::internal::Zone*) [/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Electron Framework]
[Info  - 6:55:39 AM] Connection to server got closed. Server will restart.
[Error - 6:55:39 AM] Request textDocument/hover failed.
Error: Connection got disposed.
	at Object.dispose (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/main.js:876:25)
	at Object.dispose (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-languageclient/lib/client.js:57:35)
	at LanguageClient.handleConnectionClosed (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-languageclient/lib/client.js:2036:42)
	at LanguageClient.handleConnectionClosed (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-languageclient/lib/main.js:127:15)
	at closeHandler (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-languageclient/lib/client.js:2023:18)
	at CallbackList.invoke (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/events.js:120:36)
	at closeHandler (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/main.js:226:26)
	at CallbackList.invoke (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/events.js:62:39)
	at Emitter.fire (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/events.js:120:36)
	at IPCMessageReader.fireClose (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/messageReader.js:111:27)
	at ChildProcess.<anonymous> (/Users/rikki/Desktop/cicero-vscode-extension/client/node_modules/vscode-jsonrpc/lib/messageReader.js:213:45)
	at ChildProcess.emit (events.js:208:15)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:430:11)
	at Socket.emit (events.js:203:13)
	at Pipe.<anonymous> (net.js:588:12)`


The promise/async mashup was my mistake I will rectify it.
I was going for a neat and quick fix with findUp but yes sir I will try to fix it with the existing sync code. 
Also, I had talked to @jeromesimeon about the hover feature too, if you could look into it and tell me if its something that the extension can make use for that would be great.